### DISCUSSION TITLE
Scrape metrics from multiple resource groups - selected RG not showing up

### DISCUSSION THREAD

**Discussion by ayouba88** [tomkerkhove/promitor#1731] [Q&A]
Hello, 

I have an azure account with multiple subscribers, and each subscriber with multiple groups.
I have confguired a seperate scraper/discovery container per subscribers, and in the metrics file i confguired only one group for each subscriber.
tfter reviewing the data/metrics i see that the promitor is scraping from SEVERAL RGs, but it not scraping from the RG that i have specified in the metrics file

this is the RG that i have confguired --> rg-%%%%%%-prod-%%%-management-atomiqtools

as you can see below, the promitor is scraping from several groups but not the group that i have confguired 
**all the groups are under same subscription

curl http://-----------:8900/metrics
```
# HELP active_connections Active Connections
# TYPE active_connections gauge
active_connections{resource_group="rg-%%%%%%-prod-%%%-management-atomiqtools",subscription_id="-----------------------------------------",resource_uri="subscriptions/-----------------------------------------/resourceGroups/rg-%%%%%%-prod-%%%-management-atomiqtools/providers/Microsoft.DBforPostgreSQL/servers/testsqldb99",instance_name="testsqldb99"} 4 1629213482831
active_connections{resource_group="rg-%%%%%%-prod-%%%-management-atomiqtools",subscription_id="-----------------------------------------",resource_uri="subscriptions/-----------------------------------------/resourceGroups/rg-%%%%%%-prod-%%%-management-atomiqtools/providers/Microsoft.DBforPostgreSQL/servers/psql-atomiq-finops-%%%mgmt",instance_name="psql-atomiq-finops-%%%mgmt"} 4 1629213480889
# HELP Availability The percentage of availability for the storage service or the specified API operation. Availability is calculated by taking the TotalBillableRequests value and dividing it by the number of applicable requests, including those that produced unexpected errors. All unexpected errors result in reduced availability for the storage service or the specified API operation.
# TYPE Availability gauge
Availability{resource_group="rg-%%%%%%-prod-%%%-management-arcsight",subscription_id="-----------------------------------------",resource_uri="subscriptions/-----------------------------------------/resourceGroups/rg-%%%%%%-prod-%%%-management-arcsight/providers/Microsoft.Storage/storageAccounts/%%%mgmtarcsightlogger",instance_name="%%%mgmtarcsightlogger"} NaN 1629213427767
Availability{resource_group="rg-%%%%%%-prod-%%%-management-devopsstorage1",subscription_id="-----------------------------------------",resource_uri="subscriptions/-----------------------------------------/resourceGroups/rg-%%%%%%-prod-%%%-management-devopsstorage1/providers/Microsoft.Storage/storageAccounts/%%%mgmtfilestorage01",instance_name="%%%mgmtfilestorage01"} NaN 1629213427907
# HELP azure_storage_accounts_used_capacity_average The amount of storage used by the storage account. For standard storage accounts, it's the sum of capacity used by blob, table, file, and queue. For premium storage accounts and Blob storage accounts, it is the same as BlobCapacity or FileCapacity.
# TYPE azure_storage_accounts_used_capacity_average gauge
azure_storage_accounts_used_capacity_average{resource_group="rg-%%%%%%-prod-%%%-management-arcsight",subscription_id="-----------------------------------------",resource_uri="subscriptions/-----------------------------------------/resourceGroups/rg-%%%%%%-prod-%%%-management-arcsight/providers/Microsoft.Storage/storageAccounts/%%%mgmtarcsightlogger",instance_name="%%%mgmtarcsightlogger"} 400 1629213429086
azure_storage_accounts_used_capacity_average{resource_group="rg-%%%%%%-prod-%%%-management-devopsstorage1",subscription_id="-----------------------------------------",resource_uri="subscriptions/-----------------------------------------/resourceGroups/rg-%%%%%%-prod-%%%-management-devopsstorage1/providers/Microsoft.Storage/storageAccounts/%%%mgmtfilestorage01",instance_name="%%%mgmtfilestorage01"} 232262595296 1629213428987
# HELP azur_virtual_machines_percentage_cpu_average Average percentage cpu usage on an Azure virtual machine
# TYPE azur_virtual_machines_percentage_cpu_average gauge
azur_virtual_machines_percentage_cpu_average{resource_group="rg-%%%%%%-prod-%%%-management-atomiqtools",subscription_id="-----------------------------------------",resource_uri="subscriptions/-----------------------------------------/resourceGroups/rg-%%%%%%-prod-%%%-management-atomiqtools/providers/Microsoft.Compute/virtualMachines/atomiqmonitoringcatalog02-vm",instance_name="atomiqmonitoringcatalog02-vm"} 7.6295 1629213480992
azur_virtual_machines_percentage_cpu_average{resource_group="rg-%%%%%%-prod-%%%-management-arcsight",subscription_id="-----------------------------------------",resource_uri="subscriptions/-----------------------------------------/resourceGroups/rg-%%%%%%-prod-%%%-management-arcsight/providers/Microsoft.Compute/virtualMachines/%%%mgmt-arc",instance_name="%%%mgmt-arc"} 1.1835 1629213421146
azur_virtual_machines_percentage_cpu_average{resource_group="rg-%%%%%%-prod-%%%-management-identityvmgroup",subscription_id="-----------------------------------------",resource_uri="subscriptions/-----------------------------------------/resourceGroups/rg-%%%%%%-prod-%%%-management-identityvmgroup/providers/Microsoft.Compute/virtualMachines/idmsrv-01",instance_name="idmsrv-01"} 0.55325 1629213425931
```

could you please suggest what can be the root cause ? 

Thanks and best regards

**Comment by tomkerkhove** [comment]
Hi, would it be possible to share your metrics declaration and resource discovery definition?

**Comment by ayouba88** [comment]


**Reply by tomkerkhove** [reply]
So your question is, why do you see resources from resource groups other than `rg-smartops-prod-att-management-atomiqtools`, is that correct?

If so, `azureMetadata.resourceGroupName` is the default resource group name for *declared resources*, not *discovered resources*. If you want to limit it, then you have to add the restriction in the discovery.

**Reply by tomkerkhove** [reply]
Created https://github.com/tomkerkhove/promitor/issues/1735

**Comment by ayouba88** [comment]
Thanks for the explanation, i will check this. but the most critical issue for me is that the group that i do configure do not show up, while the rest of the groups do show up with metrics
so i don't care if i see the rest of the groups, but the issue i am not able to get the one i need, could it be that i did filter it out in someway ? 


Best regards, 
Ayoub

**Reply by tomkerkhove** [reply]
Does the scraper identity have the same permissions as the resource discovery one in terms of the scope? Do they both have access to that resource?

**Reply by ayouba88** [reply]
hi,

yes, the scraper identity have same permissions, i have for the SP full read permissions with reader 
role on the resource
the scraper have the monitoring reader role assigned and the discovery have reader role assigned

some additional data 
in the discovery conf file i have the default postgresDB configurations
- name: postgres-databases
  type: PostgreSql

and for the scraper i have the below  configurations for example
- azureMetricConfiguration:
    aggregation:
      type: Average
    metricName: backup_storage_used
  description: Backup Storage Used
  name: backup_storage_used
  resourceType: PostgreSql
  resourceDiscoveryGroups:
  - name: postgres-databases



![image](https://user-images.githubusercontent.com/63043760/130574451-8821da35-7cdc-4026-a271-0a267f0bb778.png)

but in the logs of the scraper i see no values from postgresDB at all

**Reply by tomkerkhove** [reply]
Does the resource discovery agent discover them? You can verify by calling the API (`api/v1/docs`).

**Comment by ayouba88** [comment]
hi,

if i use the Discovery API, then i get the following response for the resource discovery group of postgress 
port 9901 is my discovery port

i tried 
```
API --> <myhost>:9901/api/v1/resources/groups/postgres-databases/discover
response --> page not found 
```

so i tried the below 
```
API --> <myhost>:9901/api/v1/resources/groups/PostgreSql/discover
response --> {"information":"No resource discovery group was found with specified name"}

API --> <myhost>:9901/api/v1/resources/groups/rg-smartops-prod-%%%%%%-atomiqtools/discover
response --> {"information":"No resource discovery group was found with specified name"}
```

if i check all the groups that i have using 
```
API --> <myhost>:9901/api/v1/resources/groups/
response -->
[{"name":"service-bus-landscape","type":"ServiceBusNamespace","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"api-gateways","type":"ApiManagement","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"app-plan-landscape","type":"AppPlan","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"automation-accounts","type":"AutomationAccount","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"container-instances","type":"ContainerInstance","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"container-registry-landscape","type":"ContainerRegistry","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"cosmos-accounts","type":"CosmosDb","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"dps","type":"DeviceProvisioningService","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"event-hubs-landscape","type":"EventHubs","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"front-door-landscape","type":"FrontDoor","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"iot-gateways","type":"IoTHub","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"key-vaults","type":"KeyVault","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"network-interfaces","type":"NetworkInterface","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"postgres-databases","type":"PostgreSql","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"redis-caches","type":"RedisCache","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"sql-databases","type":"SqlDatabase","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"sql-managed-instances","type":"SqlManagedInstance","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"sql-elastic-pools","type":"SqlElasticPool","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"sql-servers","type":"SqlServer","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"storage-accounts","type":"StorageAccount","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"synapse-apache-spark-pools","type":"SynapseApacheSparkPool","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"synapse-sql-pools","type":"SynapseSqlPool","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"synapse-workspaces","type":"SynapseWorkspace","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"virtual-machines","type":"VirtualMachine","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"virtual-machine-scale-sets","type":"VirtualMachineScaleSet","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}},{"name":"web-apps-north-europe","type":"WebApp","criteria":{"include":{"tags":{},"subscriptions":[],"resourceGroups":[],"regions":[]}}}]
```

**Reply by tomkerkhove** [reply]
As you can see in the last response, the resource discovery groups that you are looking for (`PostgreSql` or `rg-smartops-prod-att-management-atomiqtools`) do not exist in your [resource discovery group declaration](https://docs.promitor.io/configuration/v2.x/resource-discovery) so that's normal.

When calling `postgres-databases`, does it return a page not found or an HTTP 404? Can you try calling it through Postman?

**Reply by ayouba88** [reply]
i can't use postman on the network there, but if i use the browser i get error 500 

![image](https://user-images.githubusercontent.com/63043760/131465727-c37c4370-f5df-44fc-a9a6-f549a24f6765.png)

**Reply by tomkerkhove** [reply]
A-ha, a 500 is a different story. Did you check the logs of the agent?

**Reply by ayouba88** [reply]
which agent? resource discovery container you mean?

**Reply by tomkerkhove** [reply]
Yes

**Reply by tomkerkhove** [reply]
I see you are being throttled, can you:
- Remove the resource discovery groups that you don't need
- Change the verbosity of the logs to no longer run in debug

**Comment by ayouba88** [comment]


**Reply by tomkerkhove** [reply]
Did you try calling the endpoint after that?

**Reply by ayouba88** [reply]
I am able to call the endpoint for this group now, but still no data recieved - see below.
could it be premession issue for this RG? 

curl -v http://10.88.195.11:9901/api/v1/resources/groups/postgres-databases/discover
* About to connect() to 10.88.195.11 port 9901 (#0)
*   Trying 10.88.195.11...
* Connected to 10.88.195.11 (10.88.195.11) port 9901 (#0)
> GET /api/v1/resources/groups/postgres-databases/discover HTTP/1.1
> User-Agent: curl/7.29.0
> Host: 10.88.195.11:9901
> Accept: */*
>
< HTTP/1.1 200 OK
< Date: Wed, 01 Sep 2021 06:11:34 GMT
< Content-Type: application/json
< Content-Length: 2
< X-Version: 0.2.0
< X-Transaction-ID: 2be9ca20-1ef0-405b-a0f4-0c74ddb17db8

**Reply by tomkerkhove** [reply]
If it's returning an OK without results then it's because the identity that Promitor uses can't find any resources.

What role in Azure was assigned to the identity?

**Comment by ayouba88** [comment]
we have this role assigned for scraper 
<img width="800" alt="Screenshot_5" src="https://user-images.githubusercontent.com/63043760/136372527-326b17f6-cc93-4503-ba76-7a8caacf41a4.png">



and READER role for discovery
<img width="861" alt="Screenshot_6" src="https://user-images.githubusercontent.com/63043760/136372538-3f39a5f4-abad-40c6-9eef-27cbe694a6bd.png">

**Reply by tomkerkhove** [reply]
That should be good enough - What is the scope of it?

**Comment by ayouba88** [comment]
the scope is the the same resource, menaing the subscriber - smartops-prod-%%%%-appprod
and this subscriber have postgress resources

**Reply by tomkerkhove** [reply]
Odd.

Can you run the following query in [Azure Resource Graph](https://docs.microsoft.com/en-us/azure/governance/resource-graph/first-query-portal#run-your-first-resource-graph-query) please?
```
resources
| where type == "microsoft.dbforpostgresql/servers"
```

**Reply by ayouba88** [reply]
I do get some results but the subscriber that i am working with is not among them,
could that be a direction to the root casue?

**Reply by tomkerkhove** [reply]
That's odd indeed and can be related. I presume you have access to them as well then, right?

**Reply by ayouba88** [reply]
i have access to some of them

**Reply by tomkerkhove** [reply]
Can you run this query please?
```
resources
| where name == "<name>"
``` 

Don't forget to replace `<name>` with one of the names of the resources that you want to discover.

Once you have exported the results, can you upload it on https://www.dropbox.com/request/KwHRPNtHWUUrr7Ux1Jvg please?

**Reply by ayouba88** [reply]
hi,

i did try to run the query with the resources that i need, but i got no results. i tried the subscriber id/RG/resource ... nothing
while if i try a different subscriber i do get results

**Reply by tomkerkhove** [reply]
With subscriber you mean a subscription, I presume? Are they different kinds of subscriptions?

**Comment by ayouba88** [comment]
hi Tom, 

back on checking this issue 🔢  :)
so attached the results from the query that you sent me in the last comment

in addition, question which was raised during our checks, does promitor support flexible server metric ( postgress in our case ). my assumption was that it does, but i couldn't find anything in GitHub repo

And in case that Promitor v2.0 does not support flexible server metric, is there a plan to support it in the near future?

Thanks


[AzureResourceGraphFormattedResults-Query 2 - masked (002).csv](https://github.com/tomkerkhove/promitor/files/7299074/AzureResourceGraphFormattedResults-Query.2.-.masked.002.csv)
